change the decision variable name

hongkai-dai opened the pull request:
Now when we add decision variables to mathematical program, there are two options to specify the names of the newly added variables.

1. Specify each variable name explicitly, as
   ```
   AddContinuousVariable(rows, cols, std::vector<string> names);
   ```
   where `names` should be a vector of size `rows * cols`, containing the name of each new variable.
2. Specify the shared name of all newly added decision variables. For each individual decision variable, the program appends an index to its name.
   ```
   AddContinuousDecisionVariables(2, 3, "X")
   ```
   will add a matrix of decision variables, with names
   ```
   X(0, 0), X(0, 1), X(0, 2)
   X(1, 0), X(1, 1), X(1, 2)
   ```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4362)
<!-- Reviewable:end -->

hongkai-dai commented:
+@siyuanfeng-tri for feature review since we talked about this before.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYBfEmO3QliNIKmK0U3:b-f597j5)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 4 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/decision_variable.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBhiwU3xE6dkJ5Iruw:-KYBhiwU3xE6dkJ5Irux:brczdal) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.cc#L16)):*
> ```cpp
> 
> std::ostream& operator<<(std::ostream& os, const DecisionVariableScalar& var) {
>   os << var.name();
> ```

BTW, maybe print value?

---

*[drake/solvers/decision_variable.h, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBk_TWBZCSq1hQdyoI:-KYBk_TWBZCSq1hQdyoJ:b-fmyzfd) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.h#L130)):*
> ```objc
> template <>
> struct NumTraits<drake::solvers::DecisionVariableScalar> {
>   static inline int digits10() { return 0; }
> ```

i don't quite get this part. 

---

*[drake/solvers/mathematical_program.h, line 686 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBjwEN2pa6JeSi1eRS:-KYBjwEN2pa6JeSi1eRT:b-i9tle3) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L686)):*
> ```objc
>    * Notice that the (i,j)'th entry and (j,i)'th entry has the same name.
>    */
>   DecisionVariableMatrixX AddSymmetricContinuousVariables(
> ```

BTW, i am not very good at naming stuff either, maybe Sym as default? something a bit more descriptive. same for the others? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/decision_variable.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBhiwU3xE6dkJ5Iruw:-KYBlcpLYebVzhGui6ww:b-u8n0q5) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.cc#L16)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, maybe print value?
</blockquote></details>

The value pointer could be nullptr, if the decision variable is not created from a `MathematicalProgram` object. For example, the following code would run into a segmentation fault if we print the value
```
Eigen::Matrix<DecisionVariableScalar, 2, 2> X;
cout << X
```
Besides, the value is meaningless until we call the `Solve` function of `MathematicalProgram`, so I decide not to print it.

---

*[drake/solvers/decision_variable.h, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBk_TWBZCSq1hQdyoI:-KYBmAgjm_H2YUQIqA7b:b-f1jr) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.h#L130)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i don't quite get this part. 
</blockquote></details>

This is for printing the `DecisionVariableScalar` in an Eigen Matrix. Eigen requires to define this trait. It is also defined in `symbolic::Expression`.

---

*[drake/solvers/mathematical_program.h, line 686 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBjwEN2pa6JeSi1eRS:-KYBmRTuuzBcZsbaTY6c:bcnkzzi) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L686)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, i am not very good at naming stuff either, maybe Sym as default? something a bit more descriptive. same for the others? 
</blockquote></details>

Hmm, Sym can also mean symbolic. I am not very sure about the name. 

Actually I am more inclined to remove the default name, and make it mandatory for the user to specify a name. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/solvers/mathematical_program.h, line 686 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBjwEN2pa6JeSi1eRS:-KYBn_yz2lNUFgdC5B6z:b-oncjgl) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L686)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Hmm, Sym can also mean symbolic. I am not very sure about the name. 

Actually I am more inclined to remove the default name, and make it mandatory for the user to specify a name. What do you think?
</blockquote></details>

ah, that's good point.. 
i definitely don't want to make it mandatory. maybe just go with symmetric? i can see people use names like b and S, for something else, and they might get confused. i don't think anyone will deliberately name their thing symmetric. are they are costs for the names to be long? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/decision_variable.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBhiwU3xE6dkJ5Iruw:-KYBo32UYLzQ7L8a_x40:bhpqbtp) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.cc#L16)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

The value pointer could be nullptr, if the decision variable is not created from a `MathematicalProgram` object. For example, the following code would run into a segmentation fault if we print the value
```
Eigen::Matrix<DecisionVariableScalar, 2, 2> X;
cout << X
```
Besides, the value is meaningless until we call the `Solve` function of `MathematicalProgram`, so I decide not to print it.
</blockquote></details>

alright. i am happy as long as there is an easy way to print name + value. is there?

---

*[drake/solvers/decision_variable.h, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBk_TWBZCSq1hQdyoI:-KYBoJSAslda1iwRNPnh:b-x4g4nb) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.h#L130)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This is for printing the `DecisionVariableScalar` in an Eigen Matrix. Eigen requires to define this trait. It is also defined in `symbolic::Expression`.
</blockquote></details>

i see. what about the change for IsSigned? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I left some minor comments.

---

Reviewed 4 of 5 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 429 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBoiTFdUWGfX6oanSv:-KYBoiTFdUWGfX6oanSw:b-fgpgi1) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L429)):*
> ```objc
> 
>   /**
>    * Add continuous variables to this MathematicalProgram, with default name
> ```

BTW: nit: per style guide, "Add" should be "Adds".

(Same comment to all of the public API methods in this file.)
(I know this was a pre-existing​ defect; figured I'd leave a note since I noticed it.)

---

*[drake/solvers/mathematical_program.h, line 680 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBnv-TEnBnT7FfOmvb:-KYBnv-TEnBnT7FfOmvc:b-qpf6jl) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L680)):*
> ```objc
>    * understand the optimization program. The default name is "S", and each
>    * variable will be named as
>    * S(0, 0)     S(1, 0)     ... S(rows-1, 0)
> ```

BTW: Please verify that this shows up OK in the generated Doxygen​ website.

(Same comment below.)

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBpRyNPK1_1v4JvkCG:-KYBpRyNPK1_1v4JvkCH:bjfokcr) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/test/decision_variable_test.cc#L18)):*
> ```cpp
>   MathematicalProgram prog;
>   auto X1 = prog.AddContinuousVariables(2, 3, "X");
>   std::cout << X1 << std::endl;
> ```

BTW: Consider using `drake::log()->debug(...);`, which is defined in drake-distro/drake/common/text_logging.h.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYBp_pP3n5DD3nI7xwf:b-9fij1c)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/decision_variable.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBhiwU3xE6dkJ5Iruw:-KYBs1sNs8pWCm7bYoBb:b-8adtuf) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.cc#L16)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

alright. i am happy as long as there is an easy way to print name + value. is there?
</blockquote></details>

Yes, for a single `DecisionVariableScalar x`, you can print the name and value as
```
DecisionVariableScalar x;
// Solve the program.
std::cout<< x; // print the name. You can also print the name by std::cout<<x.name();
std::cout<< x.value(); // print the value. The user should be responsible for guaranteeing that the value pointer is not nullptr, and the value is meaningful.
```
For an Eigen::Matrix containing `DecisionVariableScalar`, you can print the name and value as
```
auto X = prog.AddContinuousVariable(2, 3, "X"); // return a 2 x 3 Eigen::Matrix containing DecisionVariableScalar objects.
// Solve the program.
std::cout<<X; // print the name
std::cout<<GetSolution(X); // print the value as an Eigen::MatrixXd.
```

---

*[drake/solvers/decision_variable.h, line 130 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBk_TWBZCSq1hQdyoI:-KYBqs02M4ioRUelhzMs:b-9wzb3f) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/decision_variable.h#L130)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i see. what about the change for IsSigned? 
</blockquote></details>

That was a mistake in the previous commit. I was thinking DecisionVariableScalar is more like just to record the index of the variable. But it should be similar to `symbolic::Expression`. So I changed the traits to be the same as `symbolic::Expression`.

---

*[drake/solvers/mathematical_program.h, line 429 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBoiTFdUWGfX6oanSv:-KYBrbyPrS3uA76mzSwC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L429)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: per style guide, "Add" should be "Adds".

(Same comment to all of the public API methods in this file.)
(I know this was a pre-existing​ defect; figured I'd leave a note since I noticed it.)
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 680 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBnv-TEnBnT7FfOmvb:-KYBuE6Eus-T0GTYpEEJ:bwc1hwj) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L680)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: Please verify that this shows up OK in the generated Doxygen​ website.

(Same comment below.)
</blockquote></details>

Done. Good call. Thanks

---

*[drake/solvers/mathematical_program.h, line 686 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBjwEN2pa6JeSi1eRS:-KYBuH4vCikAaKxgCvnR:b-9pjy73) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/mathematical_program.h#L686)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

ah, that's good point.. 
i definitely don't want to make it mandatory. maybe just go with symmetric? i can see people use names like b and S, for something else, and they might get confused. i don't think anyone will deliberately name their thing symmetric. are they are costs for the names to be long? 

</blockquote></details>

Done. I do not think we should be too concerned about that cost at this stage.

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBpRyNPK1_1v4JvkCG:-KYBugi9itTuZaNDyb5w:b-qs9n21) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/test/decision_variable_test.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: Consider using `drake::log()->debug(...);`, which is defined in drake-distro/drake/common/text_logging.h.
</blockquote></details>

This print is meant to test if `std::cout` works for an Eigen::Matrix containing `DecisionVariableScalar` objects. I do not think I can use `drake::log()` to print this matrix. The following code does not work
```
Eigen::Matrix<DecisionVariableScalar, 2, 3> X;
drake::log()->info(X);
```
Also this code does not work either
```
drake::log()->info(Eigen::Matrix2d::Zero());
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
thanks hongkai!

---

Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYBxzqZuIWgfBmHaGqz:b-ajy2gw)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@sherm1 for platform review since it is almost Monday.

---

Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYByQf2cNMBtusk_-HC:bco7n5x)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBpRyNPK1_1v4JvkCG:-KYC6XdiruUAqbsxmXn3:bb0g6xu) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/test/decision_variable_test.cc#L18)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This print is meant to test if `std::cout` works for an Eigen::Matrix containing `DecisionVariableScalar` objects. I do not think I can use `drake::log()` to print this matrix. The following code does not work
```
Eigen::Matrix<DecisionVariableScalar, 2, 3> X;
drake::log()->info(X);
```
Also this code does not work either
```
drake::log()->info(Eigen::Matrix2d::Zero());
```
</blockquote></details>

OK. I see. We need to test the "streaming" operator `<<`.  I believe the following should work:

```
std::stringstream msg_buff;
msg_buff << X1 << std::endl;
drake::log()->debug(msg_buff.str());
```

In fact, since this is a unit test, it would be ideal to programmatically test the resulting string:

```
EXPECT_EQ(msg_buff.str(), "ExpectedString");
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBpRyNPK1_1v4JvkCG:-KYCEnOAhyCmLCcoSHxT:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/test/decision_variable_test.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

OK. I see. We need to test the "streaming" operator `<<`.  I believe the following should work:

```
std::stringstream msg_buff;
msg_buff << X1 << std::endl;
drake::log()->debug(msg_buff.str());
```

In fact, since this is a unit test, it would be ideal to programmatically test the resulting string:

```
EXPECT_EQ(msg_buff.str(), "ExpectedString");
```
</blockquote></details>

Done. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBpRyNPK1_1v4JvkCG:-KYDdl9duYd3dHSG5OUt:b-tp2ma5) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/test/decision_variable_test.cc#L18)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. Thanks!
</blockquote></details>

Great! I just wanted to confirm: Are you OK just verifying that the `<<` compiles and runs without checking the string it actually produces?

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYDe7rMKRlMgf79Zh-e:-KYDe7rMKRlMgf79Zh-f:b-mu8n0v) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/test/decision_variable_test.cc#L18)):*
> ```cpp
> GTEST_TEST(TestDecisionVariable, TestDecisionVariableValue) {
>   MathematicalProgram prog;
>   auto X1 = prog.AddContinuousVariables(2, 3, "X");
> ```

BTW, since (I believe) `X1` is the device under test, I think using `auto` here overly hides the variable's type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review: a few small things, PTAL.

---

Reviewed 2 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/decision_variable.h, line 95 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEnxs2lmmz0OGFfNBz:-KYEnxs2lmmz0OGFfNC-:b-e6rasq) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/decision_variable.h#L95)):*
> ```objc
>   friend class MathematicalProgram;
> 
>   friend std::ostream& operator<<(std::ostream& os,
> ```

This shouldn't be a friend since it only uses `var.name()` which is public. Just declare it as a free function in this header and use doxygen's `@relates` command to associate it with DecisionVariableScalar in the documentation.

---

*[drake/solvers/mathematical_program.h, line 674 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEpfxHI-V6DZwH0TK3:-KYEpfxHI-V6DZwH0TK4:b-jll6db) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L674)):*
> ```objc
> 
>   /**
>    * Adds a symmetric matrix as decision variables to this MathematicalProgram.
> ```

BTW, doxygen brief (first sentence) should say that this is "runtime sized" or something like that to distinguish it from the other signature.

---

*[drake/solvers/mathematical_program.h, line 689 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYErdHGPVrpGWn6wdwL:-KYErdHGPVrpGWn6wdwM:b-fdzyxh) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L689)):*
> ```objc
>    */
>   DecisionVariableMatrixX AddSymmetricContinuousVariables(
>       size_t rows, const std::string& name = "symmetric");
> ```

BTW, I am fine with this choice as is. In case you didn't already, consider capitalizing since (a) matrix symbols are conventionally capitalized and (b) you capitalized "X" above.

---

*[drake/solvers/mathematical_program.h, line 692 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEpnJka2_zrPGWQDRw:-KYEpnJka2_zrPGWQDRx:bnli9g9) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L692)):*
> ```objc
> 
>   /**
>    * Adds a symmetric matrix as decision variables to this MathematicalProgram.
> ```

BTW, doxygen brief (first sentence) should say that this is "compile time sized" or something like that to distinguish it from the other signature.

---

*[drake/solvers/test/decision_variable_test.cc, line 21 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEqpZVWy3WKvI5fV_6:-KYEqpZWjJRms0YGCxxa:b-dpvq6l) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/test/decision_variable_test.cc#L21)):*
> ```cpp
>   std::stringstream msg_buff1;
>   msg_buff1 << X1 << std::endl;
>   drake::log()->debug(msg_buff1.str());
> ```

Are these log outputs leftover from visual inspection of results during development? For a unit test you should instead check programmatically that msg_buff1 contains the expected result.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYEs9K8zJR6HKTCK1xG:b5dq78g)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/decision_variable.h, line 95 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEnxs2lmmz0OGFfNBz:-KYF2FPmc5fbUxLSUgtn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/decision_variable.h#L95)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This shouldn't be a friend since it only uses `var.name()` which is public. Just declare it as a free function in this header and use doxygen's `@relates` command to associate it with DecisionVariableScalar in the documentation.
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 674 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEpfxHI-V6DZwH0TK3:-KYF2cQsnXqY5zsZwTTi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L674)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, doxygen brief (first sentence) should say that this is "runtime sized" or something like that to distinguish it from the other signature.
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 689 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYErdHGPVrpGWn6wdwL:-KYF2sNSE5UMIKLZYQfb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L689)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, I am fine with this choice as is. In case you didn't already, consider capitalizing since (a) matrix symbols are conventionally capitalized and (b) you capitalized "X" above.
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 692 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEpnJka2_zrPGWQDRw:-KYF2xmrsdJd6D8u5hGm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L692)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, doxygen brief (first sentence) should say that this is "compile time sized" or something like that to distinguish it from the other signature.
</blockquote></details>

Done.

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYBpRyNPK1_1v4JvkCG:-KYF3qc6mP50RnkE_i5O:b-3p9auh) ([raw file](https://github.com/robotlocomotion/drake/blob/aa22c85bdcce38c6cbedba1b053edbe01f0b9019/drake/solvers/test/decision_variable_test.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Great! I just wanted to confirm: Are you OK just verifying that the `<<` compiles and runs without checking the string it actually produces?
</blockquote></details>

Done. Thanks for the reminder. I added the comparison.

---

*[drake/solvers/test/decision_variable_test.cc, line 18 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYDe7rMKRlMgf79Zh-e:-KYF3QRSGDiLLruJnNCK:bo7veck) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/test/decision_variable_test.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, since (I believe) `X1` is the device under test, I think using `auto` here overly hides the variable's type.
</blockquote></details>

I actually want to test the return type. I added a line to test if the returned matrix is dynamic sized.

---

*[drake/solvers/test/decision_variable_test.cc, line 21 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEqpZVWy3WKvI5fV_6:-KYF3tf8J7dBSfVOpnBj:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/test/decision_variable_test.cc#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Are these log outputs leftover from visual inspection of results during development? For a unit test you should instead check programmatically that msg_buff1 contains the expected result.
</blockquote></details>

Done. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. A few BTW comments, PTAL, then ok to merge.

---

Reviewed 1 of 5 files at r1, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/decision_variable.h, line 118 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnA8E1EdQoKZly1Ax:-KYGnA8EEciVLuP-7oI_:b9jxzf7) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/decision_variable.h#L118)):*
> ```objc
> 
> /**
>  * @relates DecisionVariableScalar.
> ```

BTW, needs a short description of what the method does as the first sentence (so doxygen can use it as the brief for the method).

---

*[drake/solvers/mathematical_program.h, line 693 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnTrkDOCThPhmETn1:-KYGnTrk-Npm3ISgkcHy:b4x3qeh) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/mathematical_program.h#L693)):*
> ```objc
> 
>   /**
>    * @brief Adds a compile time sized symmetric matrix as decision variables.
> ```

BTW, sorry shouldn't include literal `@brief` keyword here because we run doxygen in autobrief mode (meaning it always makes the first sentence the brief). In an earlier discussion we decided to forbid the `@brief` keyword as superfluous.

---

*[drake/solvers/test/decision_variable_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnzQW44crQMuNMOvo:-KYGnzQW9SKizgMD5zpV:b-5y3wp) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/test/decision_variable_test.cc#L54)):*
> ```cpp
>   std::stringstream msg_buff4;
>   msg_buff4 << X2 << std::endl;
>   drake::log()->debug(msg_buff4.str());
> ```

BTW, do you still want to retain the debugging output here and above? Fine with me if you want it, but it may have served its purpose since there won't be a human on the receiving end after this lands. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYGoSXYC93Db_n_IS9Q:bhyucc1)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/decision_variable.h, line 118 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnA8E1EdQoKZly1Ax:-KYH857t9sNzdA88Xvvg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/decision_variable.h#L118)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, needs a short description of what the method does as the first sentence (so doxygen can use it as the brief for the method).
</blockquote></details>

Done.

---

*[drake/solvers/test/decision_variable_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnzQW44crQMuNMOvo:-KYH6cIy409QdT7wv3fC:b-53tzcr) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/test/decision_variable_test.cc#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, do you still want to retain the debugging output here and above? Fine with me if you want it, but it may have served its purpose since there won't be a human on the receiving end after this lands. 
</blockquote></details>

I think I compared `msg_buff4` in the next line, to the expected output. Will that be good?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 693 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnTrkDOCThPhmETn1:-KYH9IeM2WTiauuLCug_:b-qqb0z3) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/mathematical_program.h#L693)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, sorry shouldn't include literal `@brief` keyword here because we run doxygen in autobrief mode (meaning it always makes the first sentence the brief). In an earlier discussion we decided to forbid the `@brief` keyword as superfluous.
</blockquote></details>

Done. Thank

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Some responses -- ptal.

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 674 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEpfxHI-V6DZwH0TK3:-KYHsp16BgylQfHDccsq:b-5i49j3) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L674)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still needs the first sentence, just `@brief` should be gone -- now the whole line about runtime sizing is gone.

---

*[drake/solvers/mathematical_program.h, line 693 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnTrkDOCThPhmETn1:-KYHt0LJ4f3bCZpOZlII:bejibc7) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/mathematical_program.h#L693)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. Thank
</blockquote></details>

Still needs the first sentence, just `@brief` should be gone -- now the whole line about compile time sizing is gone.

---

*[drake/solvers/test/decision_variable_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnzQW44crQMuNMOvo:-KYHtA_mCPu4SD5g5Ydz:b-gdnfzn) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/test/decision_variable_test.cc#L54)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think I compared `msg_buff4` in the next line, to the expected output. Will that be good?
</blockquote></details>

That part is fine, I'm just asking if you *also* want to keep the `log()->debug()` line here -- that's for a human to read, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYHtRZ46zjyKw2VCche:b-ts3w6k)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 674 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYEpfxHI-V6DZwH0TK3:-KYHvJyH-b6j3LsbWeKV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6412315e1fda37610833ca0d72eca314ef9761d4/drake/solvers/mathematical_program.h#L674)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Still needs the first sentence, just `@brief` should be gone -- now the whole line about runtime sizing is gone.
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 693 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnTrkDOCThPhmETn1:-KYHvKsn5pT1y22qplJl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/mathematical_program.h#L693)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Still needs the first sentence, just `@brief` should be gone -- now the whole line about compile time sizing is gone.
</blockquote></details>

Done.

---

*[drake/solvers/test/decision_variable_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYGnzQW44crQMuNMOvo:-KYHvMybAAaoRPPLpa3F:budr8x1) ([raw file](https://github.com/robotlocomotion/drake/blob/3d0e6d710267acab03f3b7ca7ef8514630ba04e7/drake/solvers/test/decision_variable_test.cc#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

That part is fine, I'm just asking if you *also* want to keep the `log()->debug()` line here -- that's for a human to read, right?
</blockquote></details>

Done. I see, thanks. I did not know what the log() is doing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot retest this please

---

Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYKASt69TSI5NGB0skJ:bzdwe9o)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYKNB493L8FKRGCLlg6:b-3f2dwx)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I completed another round of review. PTAL.

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/decision_variable.cc, line 17 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKRAtN4dzvNteS5XIr:-KYKRAtN5U8caKH33FDP:b-s3tbtt) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/decision_variable.cc#L17)):*
> ```cpp
> std::ostream& operator<<(std::ostream& os, const DecisionVariableScalar& var) {
>   os << var.name();
>   return os;
> ```

BTW, can this method body be as simple as the following?

```
return os << var.name();
```

---

*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKOrNF5c29b_5u6cBV:bx7e0es) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
I believe it's a defect when a Doxygen comment block only defines some of the parameters (it should be all or none). Can you try generating the Doxygen website and seeing if this partial definition results in Doxygen emitting a warning?

Here's some text taken from [this website](
http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdparam):


> The existence of the parameter is checked and a warning is given if the documentation of this (or any other) parameter is missing or not present in the function declaration or definition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYKQyni9CWPy6MXOZJx:b-7uxp2s)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/decision_variable.h, line 122 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKRZ558lzP5xzar6ql:-KYKRZ55D_irBejaksK2:b-4qjwec) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/decision_variable.h#L122)):*
> ```objc
>  */
> std::ostream& operator<<(std::ostream& os,
>                                 const DecisionVariableScalar& var);
> ```

BTW, odd indentation. `clang-format`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-:-KYKVAlL2_1vp44eO-N9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 5 files at r1, 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 3 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/decision_variable.cc, line 17 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKRAtN4dzvNteS5XIr:-KYKWV2hDYg19O_hynKa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/decision_variable.cc#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, can this method body be as simple as the following?

```
return os << var.name();
```
</blockquote></details>

Done.

---

*[drake/solvers/decision_variable.h, line 122 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKRZ558lzP5xzar6ql:-KYKWfA8AFmu5wgg9b3B:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/decision_variable.h#L122)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, odd indentation. `clang-format`?
</blockquote></details>

Done.

---

*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKYH9P7e0RGo-3eoXH:b-anplct) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I believe it's a defect when a Doxygen comment block only defines some of the parameters (it should be all or none). Can you try generating the Doxygen website and seeing if this partial definition results in Doxygen emitting a warning?

Here's some text taken from [this website](
http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdparam):


> The existence of the parameter is checked and a warning is given if the documentation of this (or any other) parameter is missing or not present in the function declaration or definition.

</blockquote></details>

I think it works on my computer, here is a screenshot of the doxygen. 
![Screenshot from 2016-12-06 11:13:42.png](https://files.reviewableusercontent.io/images/reviewable/6314000/lP4DBlLxumLgOsE220gKHlex/Screenshot%20from%202016-12-06%2011%3A13%3A42.png)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKfnpAEptHdArox03-:bui2ug5) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think it works on my computer, here is a screenshot of the doxygen. 
![Screenshot from 2016-12-06 11:13:42.png](https://files.reviewableusercontent.io/images/reviewable/6314000/lP4DBlLxumLgOsE220gKHlex/Screenshot%20from%202016-12-06%2011%3A13%3A42.png)

</blockquote></details>

OK I guess it's not a defect to only provide a partial list of parameter definitions. TIL. Thanks for testing!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKlrq0-nIrVXMeksze:bb9mvv2) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

OK I guess it's not a defect to only provide a partial list of parameter definitions. TIL. Thanks for testing!
</blockquote></details>

The problem with this is that it makes doxygen issue warnings, not that it doesn't generate documentation. We don't have a "-Werror" option for doxygen but it does seem like a defect to me to generate a heading "Parameters" in the documentation that then doesn't list all the parameters!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKn3a2AfxQPkcTRU2S:bz3s6i8) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

The problem with this is that it makes doxygen issue warnings, not that it doesn't generate documentation. We don't have a "-Werror" option for doxygen but it does seem like a defect to me to generate a heading "Parameters" in the documentation that then doesn't list all the parameters!
</blockquote></details>

Yes, I agree defining _all_ of the parameters is preferable. 

FTR, I actually checked out this branch and built the Doxygen website and did not see any warnings, which is why I marked this thread as "satisfied". See screenshot below:

![Screenshot from 2016-12-06 15:22:54.png](https://files.reviewableusercontent.io/images/reviewable/1388098/bjFtILuGml0j9VVr35AyP77H/Screenshot%20from%202016-12-06%2015%3A22%3A54.png)

Any idea why the warning no longer appears?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKtMSW4Rb5kOX9zKJ3:bmg5h9n) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, I agree defining _all_ of the parameters is preferable. 

FTR, I actually checked out this branch and built the Doxygen website and did not see any warnings, which is why I marked this thread as "satisfied". See screenshot below:

![Screenshot from 2016-12-06 15:22:54.png](https://files.reviewableusercontent.io/images/reviewable/1388098/bjFtILuGml0j9VVr35AyP77H/Screenshot%20from%202016-12-06%2015%3A22%3A54.png)

Any idea why the warning no longer appears?
</blockquote></details>

I think we're running in silent mode which puts the output into doxygen_cxx.log.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKu-0q6LdD5ZH-nT6A:b-59gybx) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I think we're running in silent mode which puts the output into doxygen_cxx.log.
</blockquote></details>

You're right! After looking in `doxygen_cxx.log`, I see the following warnings:

```
/home/liang/dev/drake-distro-4/drake/solvers/mathematical_program.h:663: warning: The following parameters of drake::solvers::MathematicalProgram::AddSymmetricContinuousVariables(size_t rows, const std::vector< std::string > &names) are not documented:
  parameter 'rows'
/home/liang/dev/drake-distro-4/drake/solvers/mathematical_program.h:674: warning: The following parameters of drake::solvers::MathematicalProgram::AddSymmetricContinuousVariables(size_t rows, const std::string &name="Symmetric") are not documented:
  parameter 'rows'
```

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 678 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4362#-KYKOrNF1ppRxJ9isTz0:-KYKugFuEOTgzAdCNWyq:b5zl0n4) ([raw file](https://github.com/robotlocomotion/drake/blob/1ce7787401ca42c28a037047f8b49d577bb52743/drake/solvers/mathematical_program.h#L678)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

You're right! After looking in `doxygen_cxx.log`, I see the following warnings:

```
/home/liang/dev/drake-distro-4/drake/solvers/mathematical_program.h:663: warning: The following parameters of drake::solvers::MathematicalProgram::AddSymmetricContinuousVariables(size_t rows, const std::vector< std::string > &names) are not documented:
  parameter 'rows'
/home/liang/dev/drake-distro-4/drake/solvers/mathematical_program.h:674: warning: The following parameters of drake::solvers::MathematicalProgram::AddSymmetricContinuousVariables(size_t rows, const std::string &name="Symmetric") are not documented:
  parameter 'rows'
```

Thanks!
</blockquote></details>

OK, I am going to add the documentation in the next PR. Thanks for investigating into this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4362)*
<!-- Sent from Reviewable.io -->
